src/Donations/Repositories/DonationRepository.php

```diff
@@ -0,0 +1,256 @@
+            ->orderBy('posts.post_date', 'DESC')
+            ->getAll();
+
+
+        return array_map(static function ($donation) {
+            return DonationQueryData::fromObject($donation)->toDonation();
+        }, $donations);
+    }
+
+    /**
+     * @param  int donorId
+     *
+     * @return Donation[]
+     */
+    public function getByDonorId($donorId)
+    {
+        $donations = DB::table($this->postsTable)
+            ->select(
+                ['ID', 'id'],
+                ['post_date', 'createdAt'],
+                ['post_modified', 'updatedAt'],
+                ['post_status', 'status'],
+                ['post_parent', 'parentId']
+            )
+            ->attachMeta(
+                $this->donationMetaTable,
+                'ID',
+                'donation_id',
+                ['_give_payment_total', 'amount'],
+                ['_give_payment_currency', 'currency'],
+                ['_give_payment_gateway', 'gateway'],
+                ['_give_payment_donor_id', 'donorId'],
+                ['_give_donor_billing_first_name', 'firstName'],
+                ['_give_donor_billing_last_name', 'lastName'],
+                ['_give_payment_donor_email', 'email'],
+                ['subscription_id', 'subscriptionId']
+            )
+            ->where('post_type', 'give_payment')
+            ->whereIn('ID', function (QueryBuilder $builder) use ($donorId) {
+                $builder
+                    ->select('donation_id')
+                    ->from($this->donationMetaTable)
+                    ->where('meta_key', '_give_payment_donor_id')
+                    ->where('meta_value', $donorId);
+            })
+            ->orderBy('post_date', 'DESC')
+            ->getAll();
+
+        return array_map(static function ($donation) {
+            return DonationQueryData::fromObject($donation)->toDonation();
+        }, $donations);
+    }
+
+    /**
+     * @param  Donation  $donation
+     *
+     * @return Donation
+     * @throws Exception|DatabaseQueryException
+     */
+    public function insert(Donation $donation)
+    {
+        $date = current_datetime()->format('Y-m-d H:i:s');
+
+        DB::query('START TRANSACTION');
+
+        try {
+            DB::insert($this->postsTable, [
+                'post_date' => $date,
+                'post_date_gmt' => get_gmt_from_date($date),
+                'post_status' => $donation->status,
+                'post_type' => 'give_payment'
+            ], null);
+
+            $donationId = DB::last_insert_id();
+
+            foreach ($donation->getMeta() as $metaKey => $metaValue) {
+                DB::insert($this->donationMetaTable, [
+                    'donation_id' => $donationId,
+                    'meta_key' => $metaKey,
+                    'meta_value' => $metaValue,
+                ], null);
+            }
+        } catch (Exception $exception) {
+            DB::query('ROLLBACK');
+
+            Log::error('Failed creating a donation');
```

Comment:
@alaca can we store additional information about exception in the log?

Reply:
What do you have in mind?